### PR TITLE
galaxy: handle plus sign in user token appearing in role url

### LINKED ISSUE #45475: Ansible galaxy break on some token including + characters
##### SUMMARY

In my ansible-galaxy's `requirements.yml`, my `src:` include a + character in a auth token included in the URL. 
As part of dealing with specific URL scheme like git+https://, ansible-galaxy is doing a split('+'), which fail in my case due to splitting the string too many time. It should only split before the //, or do the same thing some other way.

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME

ansible-galaxy

##### ANSIBLE VERSION

```paste below
ansible 2.6.3
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/root/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.6/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.6.6 (r266:84292, Aug  9 2016, 06:11:56) [GCC 4.4.7 20120313 (Red Hat 4.4.7-17)]
```

##### CONFIGURATION

```paste below

```

##### OS / ENVIRONMENT

RedHat 6 or 7 or CentOS 7 running in a virtualbox VM over Windows 7 

##### STEPS TO REPRODUCE

Write a `req.yml`

```yaml
---

- name: bad.repository
  src: git+https://gitlab+deploy-token-15184:XXXXXXXXX@gitlab.com/remy-garrigue/testproject.git
```

Then `ansible-galaxy install -vvv -r req.yml`

##### EXPECTED RESULTS

The role to be installed like usual

##### ACTUAL RESULTS

```python-traceback
[root@cockpit vagrant]# ansible-galaxy install -vvv -r req.yml
ansible-galaxy 2.6.3
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/root/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.6/site-packages/ansible
  executable location = /usr/bin/ansible-galaxy
  python version = 2.6.6 (r266:84292, Aug  9 2016, 06:11:56) [GCC 4.4.7 20120313 (Red Hat 4.4.7-17)]
Using /etc/ansible/ansible.cfg as config file
Opened /root/.ansible_galaxy
ERROR! Unexpected Exception, this is probably a bug: too many values to unpack
the full traceback was:

Traceback (most recent call last):
  File "/usr/bin/ansible-galaxy", line 118, in <module>
    exit_code = cli.run()
  File "/usr/lib/python2.6/site-packages/ansible/cli/galaxy.py", line 152, in run
    self.execute()
  File "/usr/lib/python2.6/site-packages/ansible/cli/__init__.py", line 155, in execute
    fn()
  File "/usr/lib/python2.6/site-packages/ansible/cli/galaxy.py", line 335, in execute_install
    role = RoleRequirement.role_yaml_parse(role)
  File "/usr/lib/python2.6/site-packages/ansible/playbook/role/requirement.py", line 168, in role_yaml_parse
    (scm, src) = role["src"].split('+')
ValueError: too many values to unpack
```

### PR DESCRIPTION
**Pull request by Akasurde**
##### SUMMARY

Fixes: #45475

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
changelogs/fragments/45475_galaxy.yml
docs/docsite/rst/galaxy/user_guide.rst
lib/ansible/playbook/role/requirement.py
test/units/galaxy/test_role_requirements.py

### CODE REVIEW

File: docs/docsite/rst/galaxy/user_guide.rst [hunk 1]
```diff
@@ -226,7 +226,7 @@ To install a specific version of a role from Galaxy, append a comma and the valu
 
 .. code-block:: bash
 
-   $ ansible-galaxy install geerlingguy.apache,v1.0.0
+   $ ansible-galaxy install geerlingguy.apache,1.0.0
```
Comment (Akasurde) [inline docs/docsite/rst/galaxy/user_guide.rst:229]: There is no version called `v1.0.0`

File: lib/ansible/playbook/role/requirement.py [hunk 2]
```diff
@@ -105,9 +105,7 @@ def role_yaml_parse(role):
                     role["src"] = "git+" + role["src"]
 
                 if '+' in role["src"]:
-                    (scm, src) = role["src"].split('+')
-                    role["scm"] = scm
-                    role["src"] = src
+                    role["scm"], dummy, role["src"] = role["src"].partition('+')
```
Comment (samdoran) [inline lib/ansible/playbook/role/requirement.py:108]: It looks like `ansible.module_utils.urls.generic_urlparse()` may be useful for parsing the URL here.

Comment (Akasurde) [reply lib/ansible/playbook/role/requirement.py:108]: With `generic_urlparse`, there is additional data manipulation is required 

```
generic_urlparse(urlparse(role['src']))
{'scheme': 'git+https', 'netloc': 'gitlab+deploy-token-312644:_aJQ9c3HWzmRR4knBNyx@gitlab.com', 'path': '/akasurde/ansible-demo', 'params': '', 'query': '', 'fragment': '', 'username': 'gitlab+deploy-token-32644', 'password': '_aJQ9c3HWzmRR4kBNyx', 'hostname': 'gitlab.com', 'port': None}
```

I would like to keep it simple for now.

Comment (samdoran) [reply lib/ansible/playbook/role/requirement.py:108]: Ah, ok. It was worth looking at.

**Review by samdoran — APPROVED** [review]

**Pull request merged**